### PR TITLE
Adding blog post on programmatically using JDK 9 tools

### PR DESCRIPTION
**Pull request by gunnarmorling**
@Sanne Here it is :)

### CODE REVIEW

File: posts/Gunnar/2017-12-06-06-calling-jdk-tools-programmatically-on-java-9.adoc [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+---
+
+Without doubt, the module system (JPMS) is the most prominent feature of Java 9. But there are many other useful additions to the JDK, which have not been discussed as widely yet.
+
+One of them is the new `ToolProvider` SPI, which defines a uniform way to invoke all the tools coming with the JDK (e.g. `jar`, `jlink` etc.) in a programmatic way.
+E.g. imagine you'd like to create a JAR archive from within your Java application.
+Before Java 9, you had two options to do so:
+
+* Start a separate process to run the `jar` binary
+* Find out what classes are used by the `jar` command internally and invoke them directly
```
Comment (Sanne) [inline posts/Gunnar/2017-12-06-06-calling-jdk-tools-programmatically-on-java-9.adoc]: Uncertain: shouldn't it be "Find out which classes" ?

File: posts/Gunnar/2017-12-06-06-calling-jdk-tools-programmatically-on-java-9.adoc [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+Both options aren't ideal.
```
Comment (Sanne) [inline posts/Gunnar/2017-12-06-06-calling-jdk-tools-programmatically-on-java-9.adoc]: unnatural negation. "Neither option is ideal" ?

File: posts/Gunnar/2017-12-06-06-calling-jdk-tools-programmatically-on-java-9.adoc [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+The former incurs some overhead for forking an OS level process and it requires a bit of coding to locate the binary to execute in the Java home directory and to correctly capture any output of the forked process.
+The second option doesn't come with the disadvantages of forking a new process, but it requires knowledge about the implementation of the tool to execute.
+So for instance in the case of the `jar` command, you'd have to know that it uses `java.util.jar.JarOutputStream` internally in order to call it directly.
```
Comment (Sanne) [inline posts/Gunnar/2017-12-06-06-calling-jdk-tools-programmatically-on-java-9.adoc]: "it uses" ? Might be confusing as it sounds like `jar` *depends* on JarOutputStream. I take it you mean that JarOutputStream is its entry point?

File: posts/Gunnar/2017-12-06-06-calling-jdk-tools-programmatically-on-java-9.adoc [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+== The Java 9 `ToolProvider` SPI
+
+Now how can the task be solved with Java 9?
```
Comment (Sanne) [inline posts/Gunnar/2017-12-06-06-calling-jdk-tools-programmatically-on-java-9.adoc]: "Now how can" .. not sure this is a natural opening in native English

File: posts/Gunnar/2017-12-06-06-calling-jdk-tools-programmatically-on-java-9.adoc [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+Let's enter `java.util.spi.ToolProvider`, a new SPI which lets you run any JDK tool programmatically in a uniform fashion, within the current process.
+I.e. there's no overhead for process forking, and also no knowledge about the inner workings of the tools is needed, as they are executed via one generic interface.
```
Comment (Sanne) [inline posts/Gunnar/2017-12-06-06-calling-jdk-tools-programmatically-on-java-9.adoc]: minor style: s/and also/also/

Comment (Sanne) [reply posts/Gunnar/2017-12-06-06-calling-jdk-tools-programmatically-on-java-9.adoc]: s/executed via one generic interface./executed via a generic interface./
or /executed via an interface./

File: posts/Gunnar/2017-12-06-06-calling-jdk-tools-programmatically-on-java-9.adoc [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        System.err,
+        "--describe-module",
+        "--file",
+        "path/to/some/module.jar"
+);
+----
+
+So what's happening here?
+The `findFirst()` method of `ToolProvider` is used to obtain an instance of the tool called "jar".
+Similarly, you could obtain an instance of "jlink", "jmod" or any other of the JDK tools.
```
Comment (Sanne) [inline posts/Gunnar/2017-12-06-06-calling-jdk-tools-programmatically-on-java-9.adoc]: s/ or any other of the JDK tools./or any other JDK tool."

File: posts/Gunnar/2017-12-06-06-calling-jdk-tools-programmatically-on-java-9.adoc [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+Just be sure to use `java.util.spi.ToolProvider` and not `javax.tools.ToolProvider`, which already existed since Java 6 and only is used to obtain instances of the Java compiler.
```
Comment (Sanne) [inline posts/Gunnar/2017-12-06-06-calling-jdk-tools-programmatically-on-java-9.adoc]: s/and only is/and is only/

File: posts/Gunnar/2017-12-06-06-calling-jdk-tools-programmatically-on-java-9.adoc [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+Once you've got a reference to the tool, it can be executed by passing
+
```
Comment (Sanne) [inline posts/Gunnar/2017-12-06-06-calling-jdk-tools-programmatically-on-java-9.adoc]: intentional paragraph break here?

File: posts/Gunnar/2017-12-06-06-calling-jdk-tools-programmatically-on-java-9.adoc [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+* `PrintStream` (or `PrintWriter`) instances to receive the standard and error output of the launched tool; in the example we just use `System.out` and `System.err` to propagate any output to the current standard output and error streams, but of course you also could redirect the streams to a file, or just capture them in memory etc.
```
Comment (Sanne) [inline posts/Gunnar/2017-12-06-06-calling-jdk-tools-programmatically-on-java-9.adoc]: standard and error output**s**

File: posts/Gunnar/2017-12-06-06-calling-jdk-tools-programmatically-on-java-9.adoc [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+* the tool's arguments as you'd pass them on the CLI; in the example we pass the `--describe-module` and `--file` options and the path to the file for which we'd like to display the module metadata
```
Comment (Sanne) [inline posts/Gunnar/2017-12-06-06-calling-jdk-tools-programmatically-on-java-9.adoc]: the tool's arguments 

Something is missing? The sentence is beginning with a whitespace? No period at the end?

File: posts/Gunnar/2017-12-06-06-calling-jdk-tools-programmatically-on-java-9.adoc [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+
+And that's all you need run any JDK tool programmatically with Java 9, using one simple unified API and without the overhead of forking a separate process.
```
Comment (Sanne) [inline posts/Gunnar/2017-12-06-06-calling-jdk-tools-programmatically-on-java-9.adoc]: s/you need run/you need to run/

**Comment by gunnarmorling** [comment]
No, it really is using it. The entry point is another class.

**Comment by gunnarmorling** [comment]
Thanks @Sanne, addressed your remarks. Going to merge it now.

**Pull request merged**